packages/clerk-js/src/utils/captcha.ts

```diff
@@ -1,55 +1,132 @@
+   */
+  retry?: 'auto' | 'never';
+  /**
+   * When retry is set to auto, retry-interval controls the time between retry attempts in milliseconds.
+   * Value must be a positive integer less than 900000.
+   * @default 8000
+   */
+  'retry-interval'?: number;
+  /**
+   * Automatically refreshes the token when it expires.
+   * Can take auto, manual or never.
+   * @default 'auto'
+   */
+  'refresh-expired'?: 'auto' | 'manual' | 'never';
+  /**
+   * A JavaScript callback invoked upon success of the challenge.
+   * The callback is passed a token that can be validated.
+   * @param token string
+   */
+  callback?: (token: string) => void;
+  /**
+   * A JavaScript callback invoked when there is an error (e.g. network error or the challenge failed)
+   * @param errorCode string
+   */
+  'error-callback'?: (errorCode: string) => void;
+}
+
+interface Turnstile {
+  execute: (container?: string | HTMLElement | null, params?: RenderOptions) => void;
+  render: (container?: string | HTMLElement | null, params?: RenderOptions) => string;
+  remove: (widgetId: string) => void;
+  reset: (widgetId: string) => void;
 }
 
 declare global {
   export interface Window {
-    turnstile: { execute: (container?: string | HTMLElement | null, params?: RenderOptions) => void };
+    turnstile: Turnstile;
   }
 }
 
 const WIDGET_CLASSNAME = 'clerk-captcha';
 
 export async function loadCaptcha(url: string) {
   if (!window.turnstile) {
-    await loadScript(url, { defer: true });
+    try {
+      await loadScript(url, { defer: true });
+    } catch (_) {
+      // Rethrow with specific message
+      clerkFailedToLoadThirdPartyScript('Cloudflare Turnstile');
+    }
   }
   return window.turnstile;
 }
+
 export const getCaptchaToken = async (captchaOptions: { siteKey: string; scriptUrl: string }) => {
   const { siteKey: sitekey, scriptUrl } = captchaOptions;
-  let captchaToken = '';
+  let captchaToken = '',
+    id = '';
 
   const div = document.createElement('div');
   div.classList.add(WIDGET_CLASSNAME);
   document.body.appendChild(div);
+
   const captcha = await loadCaptcha(scriptUrl);
+  let retries = 0;
 
-  const handleCaptchaTokenGeneration = (): Promise<string> => {
+  const handleCaptchaTokenGeneration = (): Promise<[string, string]> => {
     return new Promise((resolve, reject) => {
-      return captcha.execute(`.${WIDGET_CLASSNAME}`, {
-        sitekey,
-        retry: 'never',
-        callback: function (token: string) {
-          resolve(token);
-        },
-        'error-callback': function (err) {
-          reject(err);
-        },
-      });
+      try {
+        const id = captcha.render(`.${WIDGET_CLASSNAME}`, {
+          sitekey,
+          retry: 'never',
+          // 'retry-interval': 500,
+          'refresh-expired': 'auto',
+          callback: function (token: string) {
+            resolve([token, id]);
+          },
+          'error-callback': function (errorCode) {
```

Comment:
Cloudflare marks some error codes as ones that should not be retried.  Can we be selective about our retries?

Can we record all errors that arise during retries and pass them with comma separation to captcha_error, instead of just the last one?

Reply:
Tbh, i'd expect upon retries to get back the same error code. So create a list of it, may not bring to much value.

But yes we can more selective about the retries. We need to manually do it for now, sadly turnstile does not give the option to.

Reply:
> Tbh, i'd expect upon retries to get back the same error code. So create a list of it, may not bring to much value

I agree, but it's easy to add the list, and it would be interesting if we see otherwise.  This is a bit of a black box to us, so let's be aggressive about logging the observability they do offer.

Reply:
Sounds good, I already pushed one commit that opts out of retries based on error codes.

I will add one for the concatinated list of errors.


Btw what's ur take on going with our own retry mechanism ? I guess after these changes we can go with turnstile's way by passing `retry-interval`